pkg/watch/manager.go

```diff
@@ -33,330 +38,337 @@ type Manager struct {
-		return false, nil
-	}
-	var a, r, c []string
-	for k := range added {
-		a = append(a, k.String())
-	}
-	for k := range removed {
-		r = append(r, k.String())
-	}
-	for k := range changed {
-		a = append(c, k.String())
-	}
-	log.Info("Watcher registry found changes and/or needs restarting", "started", started, "add", a, "remove", r, "change", c)
+func (wm *Manager) addWatch(r *Registrar, gvk schema.GroupVersionKind) error {
+	wm.watchedMux.Lock()
+	defer wm.watchedMux.Unlock()
+	return wm.doAddWatch(r, gvk)
+}
 
-	readyToAdd, err := wm.filterPendingResources(added)
-	if err != nil {
-		return false, errp.Wrap(err, "could not filter pending resources, not restarting watch manager")
-	}
+func (wm *Manager) doAddWatch(r *Registrar, gvk schema.GroupVersionKind) error {
+	// lock acquired by caller
 
-	if started && len(readyToAdd) == 0 && len(removed) == 0 && len(changed) == 0 {
-		log.Info("Only changes are pending additions; not restarting watch manager")
-		return false, nil
+	if r == nil {
+		return fmt.Errorf("nil registrar cannot watch")
 	}
 
-	newWatchedKinds := make(map[schema.GroupVersionKind]vitals)
-	for gvk, vitals := range wm.watchedKinds {
-		if _, ok := removed[gvk]; !ok {
-			if newVitals, ok := changed[gvk]; ok {
-				newWatchedKinds[gvk] = newVitals
-			} else {
-				newWatchedKinds[gvk] = vitals
-			}
-		}
-	}
+	// watchers is everyone who is *already* watching.
+	watchers := wm.watchedKinds[gvk]
 
-	filteredNewWatchedKinds, err := wm.filterPendingResources(newWatchedKinds)
-	if err != nil {
-		return false, errp.Wrap(err, "could not filter new watched kinds, not restarting watch manager")
-	}
-	if len(filteredNewWatchedKinds) != len(newWatchedKinds) {
-		var missing []string
-		for k := range newWatchedKinds {
-			if _, ok := filteredNewWatchedKinds[k]; !ok {
-				missing = append(missing, k.String())
-			}
-		}
-		log.Info("previously watched resources have gone pending, removing them from the watch list", "pending", missing)
+	// m is everyone who *wants* to watch.
+	m := wm.managedKinds.Get() // Not a deadlock but beware if assumptions change...
+	if _, ok := m[gvk]; !ok {
+		return fmt.Errorf("could not mark %+v as managed", gvk)
 	}
 
-	for gvk, vitals := range readyToAdd {
-		filteredNewWatchedKinds[gvk] = vitals
+	// Sanity
+	if !m[gvk].registrars[r] {
+		return fmt.Errorf("registrar %s not in desired watch set", r.parentName)
 	}
 
-	if err := wm.restartManager(filteredNewWatchedKinds); err != nil {
-		return false, errp.Wrap(err, "could not restart watch manager: %s")
+	if watchers.registrars[r] {
+		// Already watching.
+		return nil
 	}
 
-	wm.watchedKinds = filteredNewWatchedKinds
-	return true, nil
-}
+	// TODO(OREN) - should we support structured dynamic watches (using GetInformerForKind?)
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(gvk)
+	informer, err := wm.mgr.GetCache().GetInformer(u)
+	if err != nil || informer == nil {
+		// This is expected to fail if a CRD is unregistered.
```

Comment:
What do we do if Add is called too early for the resource? Return the error and wait on retry? This could cause some controllers to stop working properly if they get one bad resource (such as a sync config on a resource that doesn't exist). It may be worth keeping the old model here, where Add() just creates the intent and a separate goroutine on a timer scans for established resources and actually updates the informers.

This also makes a call to add non-blocking, which should make the config and ConstraintTemplate controllers this would also allow some parallelism in bootstrapping watches at startup

Reply:
> What do we do if Add is called too early for the resource? Return the error and wait on retry? This could cause some controllers to stop working properly if they get one bad resource (such as a sync config on a resource that doesn't exist). 
That was the plan. I don't _think_ that could disable a controller, as controller-runtime uses a  rate limiting work queue with per-item rate limiting - other requests should get a chance while this one goes into backoff.

There is another way things could get stuck - if a gvk is known but then `List` fails continuously - `GetInformer()` does a `WaitForCacheSync()`. It sounds like even with the previous model, I'd still need to run `GetInformer()` in a separate goroutine.

+1 on non-blocking and adding parallelism for startup.

Reply:
@maxsmythe I started sketching out how a reconcile-based approach [might look](https://github.com/shomron/gatekeeper/blob/dynamic-cache-reconcile/pkg/watch/manager.go#L198-L326), and wasn't happy with how complex it was turning out. Ultimately I opted for adding a non-blocking variant of `GetInformer` called  [`GetInformerNonBlocking`](https://github.com/shomron/gatekeeper/blob/3142f18f09d0fb31a8fdac7315868b7de4c692bb/pkg/dynamiccache/informer_cache.go#L144-L155). Please take a look and let me know if that addresses your concerns above.